akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -307,13 +304,21 @@ private[http] object HttpServerBluePrint {
     val timeout:       Duration,
     val handler:       HttpRequest ⇒ HttpResponse)
 
-  private class TimeoutAccessImpl(request: HttpRequest, initialTimeout: FiniteDuration, requestEnd: Future[Unit],
+  private class TimeoutAccessImpl(request: HttpRequest, initialTimeout: Duration, requestEnd: Future[Unit],
                                   trigger: AsyncCallback[(TimeoutAccess, HttpResponse)], materializer: Materializer)
     extends AtomicReference[Future[TimeoutSetup]] with TimeoutAccess with (HttpRequest ⇒ HttpResponse) { self ⇒
     import materializer.executionContext
 
-    set {
-      requestEnd.fast.map(_ ⇒ new TimeoutSetup(Deadline.now, schedule(initialTimeout, this), initialTimeout, this))
+    initialTimeout match {
+      case timeout: FiniteDuration ⇒ set {
+        requestEnd.fast.map(_ ⇒ new TimeoutSetup(Deadline.now, schedule(timeout, this), timeout, this))
+      }
+      case _ ⇒ set {
+        requestEnd.fast.map(_ ⇒ new TimeoutSetup(Deadline.now, new Cancellable {
+          override def isCancelled: Boolean = true
+          override def cancel(): Boolean = true
+        }, Duration.Inf, this))
```

Comment:
Could this `Cancellable` be a singleton object instead to save the allocation on each request?